spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/ReactorSleuth.java

```diff
@@ -55,69 +52,77 @@ private ReactorSleuth() {
+	// instrumentation injected both BeanFactory and also
+	// ConfigurableApplicationContext. This chooses the more narrow
+	// signature as it is simpler than explaining instanceof checks.
 	public static <T> Function<? super Publisher<T>, ? extends Publisher<T>> scopePassingSpanOperator(
-			BeanFactory beanFactory) {
+			ConfigurableApplicationContext springContext) {
 		if (log.isTraceEnabled()) {
-			log.trace("Scope passing operator [" + beanFactory + "]");
+			log.trace("Scope passing operator [" + springContext + "]");
 		}
 
+		// keep a reference outside the lambda so that any caching will be visible to
+		// all publishers
+		LazyCurrentTraceContext lazyCurrentTraceContext = new LazyCurrentTraceContext(
+				springContext);
+
 		return Operators.liftPublisher((p, sub) -> {
-			// While supply of scalar types may be deferred, we don't currently scope
-			// production of values in a trace context. This prevents excessive overhead
-			// when using constant results such as Flux/Mono #just, #empty, #error
+			// We don't scope scalar results as they happen in an instant. This prevents
+			// excessive overhead when using Flux/Mono #just, #empty, #error, etc.
 			if (p instanceof Fuseable.ScalarCallable) {
 				return sub;
 			}
 
-			if (beanFactory instanceof ConfigurableApplicationContext
-					&& ((ConfigurableApplicationContext) beanFactory).isActive()) {
-				if (log.isTraceEnabled()) {
-					log.trace("Spring Context [" + beanFactory
-							+ "] already refreshed. Creating a scope "
-							+ "passing span subscriber with Reactor Context " + "["
-							+ sub.currentContext() + "] and name [" + name(sub) + "]");
+			Context context = sub.currentContext();
+
+			if (log.isTraceEnabled()) {
+				log.trace("Spring context [" + springContext + "], Reactor context ["
+						+ context + "], name [" + name(sub) + "]");
+			}
+
+			// Check fast path, where we don't need to defer access to the trace context
+			CurrentTraceContext unwrapped = lazyCurrentTraceContext.unwrap();
+			if (unwrapped != null) {
+				TraceContext parent = traceContext(context, unwrapped);
+				if (parent == null) {
+					return sub; // no need to trace
 				}
 
-				return scopePassingSpanSubscription(beanFactory, sub);
+				if (log.isTraceEnabled()) {
+					log.trace(
+							"Creating a scope passing span subscriber with Reactor Context "
+									+ "[" + context + "] and name [" + name(sub) + "]");
+				}
+				return new ScopePassingSpanSubscriber<>(sub, context, unwrapped, parent);
 			}
 			if (log.isTraceEnabled()) {
-				log.trace("Spring Context [" + beanFactory
-						+ "] is not yet refreshed, falling back to lazy span subscriber. Reactor Context is ["
+				log.trace("Falling back to lazy span subscriber. Reactor Context is ["
 						+ sub.currentContext() + "] and name is [" + name(sub) + "]");
 			}
-			return new LazySpanSubscriber<>(
-					new SpanSubscriptionProvider<>(beanFactory, sub));
+			return new LazyScopePassingSpanSubscriber<>(sub, context,
```

Comment:
now that we no longer inject beanfactory, we never get here. the code is dead running all the tests as far as I can tell with breakpoints.. even in the case of spring-cloud-context. because the ConfigurableApplicationContext is never not active anymore. It could be a massive simplification to remove the if statement and this special case...